adafruit_midi/channel_pressure.py

```diff
@@ -0,0 +1,76 @@
+# The MIT License (MIT)
+#
+# Copyright (c) 2019 Kevin J. Walters
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+"""
+`adafruit_midi`
+================================================================================
+
+A CircuitPython helper for encoding/decoding MIDI packets over a MIDI or UART connection.
+
+
+* Author(s): Kevin J. Walters
+
+Implementation Notes
+--------------------
+
+**Hardware:**
+
+
+
+**Software and Dependencies:**
+
+* Adafruit CircuitPython firmware for the supported boards:
+  https://github.com/adafruit/circuitpython/releases
+
+"""
+
+from .midi_message import MIDIMessage
+
+__version__ = "0.0.0-auto.0"
+__repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_MIDI.git"
+
+
+class ChannelPressure(MIDIMessage):
+    """Channel Pressure MIDI message.
+
+    :param int pressure: The pressure, 0-127.
+    """
+
+    _STATUS = 0xd0
+    _STATUSMASK = 0xf0
+    LENGTH = 2
+    CHANNELMASK = 0x0f
+
+    def __init__(self, pressure):
+        self.pressure = pressure
+        if not 0 <= self.pressure <= 127:
+            raise self._EX_VALUEERROR_OOR
+
+    # channel value is mandatory
+    def as_bytes(self, channel=None):
```

Comment:
I can change this but I would like to keep thoses classes as small as possible in memory use for the M0 world.

I observe @dhalbert has a `to_bytes()` in https://github.com/adafruit/Adafruit_CircuitPython_BluefruitConnect/blob/master/adafruit_bluefruit_connect/color_packet.py which was mentioned as a reference pattern for this work.

Reply:
The essence of the issue with arguments is I wasn't sure if I should keep the interface consistent or declared identically across all objects. Some of the messages require a channel, others don't. I'll refresh my basic python knowledge on `*` and non-positional arguments tomorrow.

Reply:
In general, I think midi libraries have a dilemma here wrt channel. Does a NoteOn() object represent a just a note being on or does it represent a note being on for a particular channel or for a channel that can change - the final permutation is required if you want to send that to a specified channel and then to another. I've gone for the first approach where the note is not associated with a channel until it's converted to its wire protocol representation. For received notes I am (currently) handing the channel back with the message rather than storing it in the object.

Many applications won't be interested in the channel number on incoming messages. I'm using it for my PSG-esque synth, 1 and 2 go to the external square wave VCA setup, channel 3 is for a very coarse (due to low memory!), single volume wavetable.

Reply:
My preference would be to have the channel as an attribute on the message. That way it's easy to keep track of. When forwarding messages you can always overwrite the incoming value.
 
It's important to consider that returning multiple values is triggering a second allocation for the tuple in addition to the note itself. I'd rather have note be a little larger than doing a second allocation.

`to_bytes` should also be changed to `__bytes__`. It is how the USB descriptor works: https://github.com/adafruit/usb_descriptor/blob/master/adafruit_usb_descriptor/hid.py#L66

Reply:
Ok, I'll have time around Friday to have look into this. Just to confirm, you want an (immutable) `bytes` type rather than `bytearray`. I did it the latter way in case the user wanted to do some dirty messing around with the data for some reason. I couldn't think of any particular reason they'd want to do that, it just seemed more flexible.

Reply:
Ya, return a bytes. They can always put it back into a bytearray if they want to change it.

Reply:
This means the `MIDI.send()` method now changes the objects it's passed as it has to update the channel numbers in them to the out channel.

Reply:
It also means `MIDI.receive` changes but more importantly now that channel is held in the message it feels like the setting of it for _incoming_ messages belongs in the `from_bytes` method.

Reply:
The `from_message` class method per message now needs to see the status byte to be able to extract the channel number.

Reply:
I've just finished this. It passes tests and I've fired it up with my code and it works. It gained 894 bytes on **mpy** files and that's pushed me to another `MemoryError` and forced me to drop `ChannelPressure` `import` in my synth demo code, so no more aftertouch :(

I need to check docs to make sure I got everything that's changed.

Reply:
For comparison, I've stumbled across some other work that started before mine. This one is oriented towards MIDI files, there's a large degree of overlap but will be a bit more in the MIDI files. https://github.com/adafruit/Adafruit_Learning_System_Guides/blob/master/NeoTrellis_M4_MIDI_Synth/events.py does store the channel in the parent event object called `MIDIEvent`.

Reply:
`__bytes__` has a sting in its tail. I know you didn't state this but I had assumed I could call this as `bytes(object)`. That works ok in the unit tests on Python 3. Turns out micropython doesn't support that: https://github.com/micropython/micropython/issues/3158 plus forum discussion in https://forums.adafruit.com/viewtopic.php?f=60&t=150071